vulnerabilities/phantom-functions.md

```diff
@@ -0,0 +1,163 @@
+    uint amount,
+    uint deadline,
+    uint8 v,
+    bytes32 r,
+    bytes32 s,
+    uint toChainID
+) external {
+    address _underlying = AnyswapV1ERC20(token).underlying();
+
+    IERC20(_underlying).permit(from, address(this), amount, deadline, v, r, s);
+    TransferHelper.safeTransferFrom(_underlying, from, token, amount);
+    AnyswapV1ERC20(token).depositVault(amount, from);
+    _anySwapOut(from, token, to, amount, toChainID);
+}
+```
+
+>   For ERC-20 tokens that support it, permit is an alternative to the standard approve call: it allows an off-chain secure signature to be used to register an allowance. The permitter is approving the beneficiary to spend their money, by signing the permit request.
+
+>   The permit approach has several advantages including: there is no need for a separate transaction (spending gas) to approve a spender, allowances have a deadline, transfers can be batched and more.
+
+>   More on erc-20 permit and why it is better than the standard approve call can be found [here](https://forum.openzeppelin.com/t/why-is-erc20permit-better-than-approve-transferfrom/7478) and [here](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20Permit)
+
+
+While the above code looks normal, it takes on an abnormal nature if the contract of the underlying token contains a fallback function. A classic example of a token always having a fallback is wrapped native tokens (WETH, WBNB, WAVAX, etc)
+
+Here's how the attack scenario would look like:
+
+1. Alice, the victim user, has approved 10 WETH to the Router.
+2. Bob, the attacker, calls the `anySwapOutUnderlyingWithPermit()` function on the Router contract with an amount of 10 WETH and arbitrary v, r and s values
+
+3. Permit Function Call: The Router contract attempts to call the permit function on the WETH contract with the purpose of delegating authorization, i.e
+
+```
+IERC20(_underlying).permit(from, address(this), amount, deadline, v, r, s)
+```
+
+Since permit does not exist on WETH, the call falls through to the fallback function.
+
+4. The fallback function in WETH executes and does not revert which the Router contract mistakenly interprets as a successful permit call.
+
+5. Token Transfer: The Router contract calls `safeTransferFrom()` to move 10 WETH from Alice’s account, wrongly assuming that the permit authorization was successful.
+
+6. Funds Bridged: The Router completes the transfer process, bridging 10 WETH to Bob's address on another blockchain, resulting in Bob stealing 10 WETH from Alice.
+
+
+Assumptions made on the return value returned from WETH seems like the obvious victim here. While true, phantom functions also played a part here. Read along to see why.
+
+## Vulnerable flashloan functions
+
+This vulnerability was discovered by auditors in [code4rena](https://github.com/code-423n4/2023-06-lybra-findings/issues/769). I will reproduce the vulnerable contract in a simplified manner demonstrating how it led to financial losses.
```

Comment:
Should avoid using "I" here. Instead would be better to say something like "The vulnerable contract is reproduced below..."